aten/src/ATen/native/SoftMax.cpp

```diff
@@ -465,23 +482,62 @@ Tensor log_softmax(const Tensor& self, Dimname dim, optional<ScalarType> dtype)
   return at::log_softmax(self, dimname_to_position(self, dim), dtype);
 }
 
-Tensor masked_softmax_cpu(const Tensor& input, const Tensor& mask) {
+Tensor masked_softmax_cpu(const Tensor& input, int64_t dim, const Tensor& mask) {
   Tensor output = at::empty_like(input, input.options());
+  auto mask_ = mask.is_contiguous() ? mask : mask.contiguous();
   TORCH_CHECK(
       input.sizes() == mask.sizes(), "Mask shape should match input shape");
-  TORCH_CHECK(mask.is_contiguous(), "Mask should always be contiguous");
   TORCH_CHECK(
       mask.scalar_type() == ScalarType::Bool,
       "Mask should be a boolean tensor");
+
+  auto input_ = input.contiguous();
+  int64_t dim_ = maybe_wrap_dim(dim, input_.dim());
+
+  if (input_.dim() == 0) {
+    input_ = input_.view(1);
+  }
+
   AT_DISPATCH_FLOATING_TYPES_AND(
-      at::ScalarType::BFloat16, input.scalar_type(), "log_softmax", [&] {
+      at::ScalarType::BFloat16, input.scalar_type(), "masked_softmax", [&] {
         host_softmax<
             scalar_t,
             false /* LogSoftMax */,
             true /* MaskedSoftMax */>(
-            output, input, input.dim() - 1, mask.data_ptr<bool>());
+            output, input_, dim_, mask_.data_ptr<bool>());
       });
   return output;
 }
+
+Tensor masked_softmax_backward_cpu(
+    const Tensor& grad_,
+    const Tensor& output_,
+    int64_t dim_,
+    const Tensor& mask_) {
+  TORCH_CHECK(
+      grad_.sizes() == mask_.sizes(), "Mask shape should match grad shape");
+  TORCH_CHECK(
+      mask_.scalar_type() == ScalarType::Bool,
+      "Mask should be a boolean tensor");
+
+  int64_t dim = maybe_wrap_dim(dim_, grad_.dim());
+  auto grad = grad_.contiguous();
```

Comment:
contiguous() sometimes takes a lot of time. Please make sure it wont take too much time when the tensor is contiguous already, or else you should check if it is contiguous first.

Reply:
I don't think this should be an issue - https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorProperties.cpp#L62-L64 indicates that if it's already contiguous, it will just return itself

Reply:
`.contiguous()` is very fast when tensor is contiguous already.